src/libraries/System.Private.CoreLib/src/System/ComAwareWeakReference.cs

```diff
@@ -0,0 +1,179 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+#if FEATURE_COMINTEROP || FEATURE_COMWRAPPERS
+
+using static System.WeakReferenceHandleTags;
+
+namespace System
+{
+    internal sealed partial class ComAwareWeakReference
+    {
+        private readonly nint _weakHandle;
+        private ComInfo? _comInfo;
+
+        internal sealed class ComInfo
+        {
+            private readonly IntPtr _pComWeakRef;
+            private readonly long _wrapperId;
+
+            internal object? ResolveTarget()
+            {
+                return ComWeakRefToObject(_pComWeakRef, _wrapperId);
+            }
+
+            // see: syncblk.h
+            private const int IS_HASHCODE_BIT_NUMBER = 26;
```

Comment:
This is CoreCLR-specific. It would be better for it to live in CoreCLR-specific file.

Reply:
I will check if this can be factored out. 
This will apply to NativeAOT as well, when it supports COM weak references, but may not be ever applicable to Mono.

Reply:
> This will apply to NativeAOT as well

Actually I am not sure. I kind of assume the interop info is handled the same way, but that is not necessarily so.

Reply:
@AaronRobinsonMSFT can you comment on support for COM weak reference support in NativeAOT?

Reply:
I think support makes sense longer term - particularly as we move forward on `ComWrappers`. I mentioned support in some Not Yet Implemented behavior in this PR.

Reply:
I think we should support weak COM references at some point within the context of overall COM support in NativeAOT. 
I assume it is just a matter of dev cost and prioritization.

That is why I have added NYI call outs for NativeAOT here - so it would be clear where it will plug in.